redhat-appstudio/service-provider-integration-operator
# feat: adapting spiaccesscheck to remotesecrets; version2

### What does this PR do?
Extends SPIAccessCheck's functionality.

The CheckAccess first tries to find SPIAccessToken to construct a client and determine if a repository is accessible.
If no SPIAccessToken is found, this PR comes into play. We find all RemoteSecrets in the same namespace as SPIAccessCeck with the label `appstudio.redhat.com/sp.host` and filter this list of RemoteSecrets so that only those that satisfy these conditions are kept:
- has Data Obtained Condition
- has `secret.type == kubernetes.io/basic-auth`
- has target in the same namespace as SPIAccessCheck (and the RemoteSecret itself)

From these, we preferentially pick an RS that contains the repository name (from SPIAccessCheck ) in  `appstudio.redhat.com/sp.repo`.

We then use the RemoteSecret's Secret to construct an authenticated client and determine whether repository is accessible.

The SPIAccessCheck CRD is extended with the field:
```yaml
status:
  credentials: 
    remoteSecret: __name of remote secret__
    secret: __name of secret__
```

### Screenshot/screencast of this PR

#### GitHub
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/6567bf98-76e0-4d2f-a3aa-9c058ffa9d01)

#### Quay
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/52e6d801-2c36-44be-96b0-0a4b5b0f21b2)

#### GitLab
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/1baac6ad-357b-48fd-9a4d-9eddf3235362)


### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-564

### How to test this PR?

Deploy SPI and RemoteSecret
SPIO_IMG=quay.io/pbaran/spio:SVPI-564-spiaccesscheck

#### GitHub
1. Create Github Personal Access Token (classic) (https://github.com/settings/tokens) `repo` scope should suffice
2. Create (GitHub) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: github.com
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
3. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  password: "_your GitHub PAT_"
EOF
```
4. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://github.com/xbaran4/private"
EOF
```
5. You should see SPIAccessCheck similar to the state in the screenshot

#### Quay
1. Create Quay robot token. Make sure it has permission (`write`) to access your private registry
2. Create (Quay) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: quay.io
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
3. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  username: "pbaran+test" #replace with your robot name
  password: "_your robot token_"
EOF
```
4. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://github.com/pbaran/private-test" # replace with your private reg
EOF
```
5. You should see SPIAccessCheck similar to the state in the screenshot

#### GitLab
1. Create GitLab Personal Access Token (https://gitlab.com/-/profile/personal_access_tokens) `api` scope should suffice
2. Create (GitLab) RemoteSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: test-remote-secret
  namespace: default
  labels:
    appstudio.redhat.com/sp.host: gitlab.com
spec:
  secret:
    type: kubernetes.io/basic-auth
    generateName: secret-from-remote-
  targets:
    - namespace: default
EOF
```
3. Create UploadSecret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: test-remote-secret-secret
  labels:
    appstudio.redhat.com/upload-secret: remotesecret
  annotations:
    appstudio.redhat.com/remotesecret-name: test-remote-secret
type: kubernetes.io/basic-auth
stringData:
  password: "_your GitLab PAT_"
EOF
```
4. Create SPIAccessCheck with repoUrl referencing your private repo:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessCheck
metadata:
  name: spiaccesscheck-sample
  namespace: default
spec:
  repoUrl: "https://github.com/xbaran4/test"
EOF
```
5. You should see SPIAccessCheck similar to the state in the screenshot

```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-#
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-#
```
